File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int diameter =0;
    public int diameterOfBinaryTree(TreeNode root) {
        longestDia(root);
        return diameter;
    }
    public int longestDia(TreeNode root){
        if(root==null) return 0;
        int leftPath = longestDia(root.left);
        int rightPath = longestDia(root.right);
        diameter = Math.max(diameter, leftPath+rightPath);
        return Math.max(leftPath, rightPath)+1;
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int max_d=0;
        if(root==NULL) return 0;
        diameter(root,max_d);
        return max_d;
    }
    int diameter(TreeNode* root,int &max_d){
        if(root==NULL) return 0;
        int leftD=diameter(root->left,max_d);
        int rightD=diameter(root->right,max_d);
        max_d=max(max_d, leftD+rightD);
        return 1+max(leftD,rightD);
    }
};